certora/specs/Reverts.spec

```diff
@@ -0,0 +1,253 @@
+    if (updatedPrice != lastPrice) {
+        priceChanged = true;
+        lastPrice = updatedPrice;
+    }
+    return updatedPrice;
+}
+
+function summaryMulDivUp(uint256 x,uint256 y, uint256 d) returns uint256 {
+    // Safe require because the reference implementation would revert.
+    return require_uint256((x * y + (d-1)) / d);
+}
+
+function summaryToAssetsUp(uint256 shares, uint256 totalAssets, uint256 totalShares) returns uint256 {
+    return summaryMulDivUp(shares,
+                           require_uint256(totalAssets + (10^6)),
+                           require_uint256(totalShares + (10^6)));
+}
+
+function summaryToSharesUp(uint256 assets, uint256 totalAssets, uint256 totalShares) returns uint256 {
+    return summaryMulDivUp(assets,
+                           require_uint256(totalShares + (10^6)),
+                           require_uint256(totalAssets + (10^6)));
+}
+
+definition ORACLE_SCALE() returns uint256  = 10^36;
+
+definition exactlyOneZero(uint256 assets, uint256 shares) returns bool =
+  (assets == 0 && shares != 0) || (assets != 0 && shares == 0);
+
+definition wDivUp(uint256 x,uint256 y) returns uint256 = summaryMulDivUp(x, WAD(), y);
+
+definition wMulDown(uint256 x,uint256 y) returns uint256 = summaryMulDivDown(x, y, WAD());
+
+definition computeLinearCombination(mathint ltv, mathint lltv, mathint preLltv, mathint yAtPreLltv, mathint yAtLltv)
+    returns mathint =
+    wMulDown(wDivDown(require_uint256(ltv - preLltv),
+            require_uint256(lltv - preLltv)),
+        require_uint256(yAtLltv - yAtPreLltv)) + yAtPreLltv;
+
+// Checks that onMorphoRepay is only triggered by Morpho
+rule onMorphoRepaySenderValidation(env e, uint256 repaidAssets, bytes data) {
+    onMorphoRepay@withrevert(e, repaidAssets, data);
+    assert e.msg.sender != currentContract.MORPHO => lastReverted;
+}
+
+// Check that preLiquidate reverts when its inputs are not validated.
+rule preLiquidateInputValidation(env e, address borrower, uint256 seizedAssets, uint256 repaidShares, bytes data) {
+    requireInvariant preLltvConsistent();
+    requireInvariant preLCFConsistent();
+    requireInvariant preLIFConsistent();
+    preLiquidate@withrevert(e, borrower, seizedAssets, repaidShares, data);
+    assert !exactlyOneZero(seizedAssets, repaidShares) => lastReverted;
+}
+
+// Check that collateralQuoted == 0 would revert by failing require-statements.
+rule zeroCollateralQuotedReverts(env e, address borrower, uint256 seizedAssets, bytes data) {
+    // Market values.
+    uint256 mTotalBorrowAssets;
+    uint256 mTotalBorrowShares;
+
+    // Position values.
+    uint256 pBorrowShares;
+    uint256 pCollateral;
+
+    uint256 collateralPrice;
+
+    requireInvariant preLltvConsistent();
+
+    uint256 collateralQuoted = require_uint256(summaryMulDivDown(pCollateral, collateralPrice, ORACLE_SCALE()));
+    uint256 borrowed = require_uint256(summaryToAssetsUp(pBorrowShares, mTotalBorrowAssets, mTotalBorrowShares));
```

Comment:
I've implemented the one that actually checks the revert doing munging, do you think I should add it to repository, at the cost of adding the munging set-up.

Reply:
> do you think I should add it to repository, at the cost of adding the munging set-up

This is a separate issue, and it was decided [here](https://github.com/morpho-org/pre-liquidation/pull/37#discussion_r1793499673) that we would not bother with munging in this PR.

I was simply suggesting for this rule to remove the highlighted code, and essentially replace it with:
```solidity
    uint256 collateralQuoted;
    uint256 borrowed;
```
which would take arbitrary values for `collateralQuoted` and `borrowed`. So it would be more general, and simpler to read

Reply:
You're right. Implemented in https://github.com/morpho-org/pre-liquidation/pull/37/commits/ebdaff6e0cdf9a9030237d530d03bf9708364178.